labs/b5.md

```diff
@@ -0,0 +1,364 @@
+
+## MAC
+Media access control (MAC) addresses are a identifiers uniquely assigned to network interfaces.
+![alt text][MAC]
+
+[MAC]: https://upload.wikimedia.org/wikipedia/commons/9/94/MAC-48_Address.svg "MAC address format"
+
+Since the MAC address is unique this is often referred to as the physical address. The octets are often written in hexadecimal and delimited by colons. An example MAC address is  `00:14:22:01:23:45`. Note that the first 3 octets refer to the Organizationally Unique Identifier (OUI) which can help identify manufacturers.  Fun fact -- the `00:14:22` above is an OUI for Dell.
+
+## ARP
+Address Resolution Protocol (ARP) is a protocol used to resolve IP addresses to MAC addresses. For example let's imagine a sender, A, who has MAC `00:DE:AD:BE:EF:00`,  broadcasting a message that essentially asks "Who has IP address `42.42.42.42` please tell A at 00:DE:AD:BE:EF:00". 
+
+If a machine, B, with MAC `12:34:56:78:9a:bc` has the IP address `42.42.42.42` they send a unicast reply back to the sender with the info "`12:34:56:78:9a:bc` has `42.42.42.42`". The sender stores this information in an arp table so whenever it receives packets meant for machine B i.e. a packet with an destination IP address of `42.42.42.42` it sends the packet to MAC it received from B.
+
+## IP
+IP addresses are means of identifying devices connected to a network under Internet Protocol. There are two versions of the internet protocol IPv4 and IPv6 that have different address format. For the sake of time we will only go over IPv4, but IPv6 is certainly gaining ground and worth checking out!
+
+IPv4 addresses are expressed in CIDR format, which is comprised of 32 bits, i.e. 4 bytes, long and are delimited by a dot (.) every byte. An example IPv4 address is `127.0.0.1`. Coincidentally this address is known as the loopback address which maps to the loopback interface on your own machine. This allows network applications to communicate with one another if they are running on the same machine, in this case your machine. But why `127.0.0.1` and not `127.0.0.0` or `127.0.0.2`?
+
+The answer is that `127.0.0.1` is simply convention, but technically any address in the network block `127.0.0.0/8` is a valid loopback address. But what exactly is a network block?
+
+In IPv4 we can partition a block of addresses into a subnet. Let's take the subnet above as an example `127.0.0.0/8`. The number that comes after the slash (`/`), in this case 8, is the subnet mask. This represents how many bits are in the network address, the remaining bits identify a host within the network. In this case the network address is `127.0.0.0` and the Mask is `255.0.0.0`. So `127.0.0.1` would be the first host in the `127.0.0.0/8` network and so on and so forth.
+
```

Comment:
we might need to explain this stuff a little more. I don't imagine a lot of people will have any intuition for what this stuff means

Reply:
I agree that this is often confusing, the newer slides should help walk them through that and office hours are always there